dialects: (vector) Add for `vector.transfer_read` and `vector.transfer_write` operations

watermelonwolverine opened the pull request:
Added stubs for "vector.transfer_read" and "vector.transfer_write".
Verification is missing but that also seems to be a lot of work.

Also fixed a small bug in TensorOrMemrefOf.verify (?)

compor commented on xdsl/dialects/builtin.py:
```diff
@@ -1891,8 +1891,8 @@ def get_resolvers(
-        if isinstance(attr, VectorType) or isinstance(attr, TensorType):
-            attr = cast(VectorType[Attribute] | TensorType[Attribute], attr)
+        if isinstance(attr, MemRefType) or isinstance(attr, TensorType):
+            attr = cast(MemRefType[Attribute] | TensorType[Attribute], attr)
```
Would you mind moving this to a separate PR with a test case?
We seem to be lacking any verification testing on this type.

compor replied:
Seeing the changes, I'd still recommend doing this.

watermelonwolverine replied:
I will do that in the next few days

compor commented on xdsl/dialects/vector.py:
```diff
@@ -292,6 +302,78 @@ def get(mask_operands: list[Operation | SSAValue]) -> CreatemaskOp:
+        # TODO verify.
+
+    @staticmethod
+    def get(
```
Same as @superlopuh commented in your other PR (#3649), we have moved to using `__init__`

compor commented on xdsl/dialects/vector.py:
```diff
@@ -292,6 +302,78 @@ def get(mask_operands: list[Operation | SSAValue]) -> CreatemaskOp:
+        # TODO verify
+
+    @staticmethod
+    def get(
+        vector: SSAValue | Operation,
```
same here as above

compor commented on xdsl/dialects/vector.py:
```diff
@@ -292,6 +302,78 @@ def get(mask_operands: list[Operation | SSAValue]) -> CreatemaskOp:
+    def verify_(self):
+        assert isa(self.source.type, MemRefType[Attribute] | TensorType[Attribute])
+        assert isa(self.result.type, VectorType[Attribute])
+        # TODO verify.
+
```
Verification might be a lot of work, but maybe you could add parts of it?
Checking rank, element type and mask should not be that bad, no?

Once this is merged, could you please open an issue to track what is missing?

compor reviewed (commented):
LGTM modulo comments

watermelonwolverine commented:
- I started implementing the verification and finished most of it 
    - Only verification for cases where the element_type of source is a vector type is missing
- Problem: To fully implement verification, as it is done in MLIR, the VectorType needs to be changed first
    - It needs a bitmap (ArrayAttr[BoolAttr]) for the scalable dimensions instead of a single IntegerAttr for the number of scalable dimensions.
    - I could skip this verification step and do everything else 
    - I did some of those changes so you can see for yourself what I mean.
- I also had to add several functions to builtin classes like AffineMap and AffineExpr. 
    - Should I do a separate PR for those, too?

watermelonwolverine commented:
- Reverted changes done to VectorType.
- Added unit tests for function added to AffineExpr and AffineMap
- Fixed some bugs with verification and did some refactoring
- Disabled code parts that rely on issue #3654 to be fixed

compor commented:
You can always verify what is currently here and document what deviates (as you have done), this is fine IMHO.

Thanks for the work so far!

superlopuh commented on xdsl/dialects/vector.py:
```diff
@@ -383,6 +397,330 @@ def __init__(
+            f'"{op.name}" requires a permutation_map with input dims of the same rank as the source type'
+        )
+
+    # WJOG9GVF: TODO fix: uncomment this when 7S4F0FZA has been fixed
+    # if mask_type:
+    #     if mask_type != inferred_mask_type:
+    #         raise VerifyException(
+    #             f'"{op.name}" inferred mask type ({inferred_mask_type}) and mask operand type ({mask_type}) don\'t match'
+    #         )
```
?

watermelonwolverine replied:
Same as below. I will move it to a new pull request

superlopuh commented on xdsl/dialects/vector.py:
```diff
@@ -383,6 +397,330 @@ def __init__(
+    TODO test
+    """
+
+    # 7S4F0FZA
+    # TODO uncomment and test this once VectorType has been fixed, see issue #3654
+    # When you do this also fix all WJOG9GVF
+
+    # inverse_permutation_map = affine_map.compress_dims(
+    #     affine_map.unused_dims_bit_vector()
+    # ).inverse_permutation()
+
+    # assert inverse_permutation_map
+
+    # mask_shape = inverse_permutation_map.compose_with_values(vector_type.get_shape())
+
+    # scalable_dims = inverse_permutation_map.eval(
+    #     [1 if dim_scalable else 0 for dim_scalable in vector_type.get_scalable_dims()],
+    #     [],
+    # )
+
+    # return VectorType(
+    #     i1,
+    #     mask_shape,
+    #     [dim_scalable == 1 for dim_scalable in scalable_dims],
+    # )
```
?

watermelonwolverine replied:
The problem is issue #3654 
As soon as this is fixed the code above can be  uncommented. I can remove it but then how should I document this?

watermelonwolverine replied:
I can move it to a new pull request and link it to the issue

superlopuh replied:
Ah yes do you think you have the bandwidth to fix this also? I have a feeling that this PR could be a draft until the other points are fixed, would help calibrate how pedantic the reviewers should be when looking at this one :)

superlopuh commented on xdsl/dialects/vector.py:
```diff
@@ -383,6 +397,330 @@ def __init__(
+    vector_type: VectorType[Attribute],
+    affine_map: AffineMap,
+) -> VectorType[I1] | None:
+    """
+    TODO test
+    """
```
?

watermelonwolverine replied:
Sorry, I will clean that up

superlopuh commented on xdsl/dialects/vector.py:
```diff
@@ -383,6 +397,330 @@ def __init__(
+
+    irdl_options = [AttrSizedOperandSegments(as_property=True), ParsePropInAttrDict()]
+
+    # assembly_format = "$source `[` $indices `]` `,` $padding  ( `,` $mask^ )? attr-dict `:`  type($source) `,` type($result)"
```
commented code

watermelonwolverine replied:
I will clean that up

superlopuh commented:
Hi @watermelonwolverine! I'd love to get this in, please let me know if you'd like some help with the PR. From my understanding, these bits of code that are currently in this PR would benefit from their own mini PRs:

- [ ] is_function_of_dim + test
- [ ] compose_with_values + test
- [ ] used/unused_dims + test

Do you have time to have a go at these? I'm hoping it will be a fairly easy cherry-picking process, I'm happy to review each one very quickly :) After that, I think this PR should be ready to merge.

watermelonwolverine commented:
This PR is actually pretty much done from my side.
There are a few verifications I had to remove because they depend on #3654 being fixed. I plan to create a new pull request afterwards to add the verifications back in.
There are a few TODOs in verification and testing, but but I think those deserve to be their own PRs.

watermelonwolverine commented:
I can split this PR into smaller PRs over the next few days if it is simply too big

superlopuh commented:
Yeah that would be great, thank you

superlopuh commented:
Just opened #3947, which fixes the above issue. Would you have the time to iterate on this in the coming weeks?

watermelonwolverine commented:
Hi, I am very busy with my thesis, but I think I can still fit this in somehow. I'll update this PR once the other PR is merged. Should I still split this PR into smaller PRs?

superlopuh commented:
That would be fantastic, but also please let me know if there's another time to check in for this, there's no particular rush, it's just nice to be able to get to 0 PRs eventually.

superlopuh commented on tests/filecheck/dialects/vector/vector_ops.mlir:
```diff
@@ -10,6 +10,9 @@ builtin.module {
     "vector.maskedstore"(%0, %2, %2, %1, %6) : (memref<4x4xindex>, index, index, vector<1xi1>, vector<1xindex>) -> ()
     "vector.print"(%6) : (vector<1xindex>) -> ()
     %7 = "vector.create_mask"(%2) : (index) -> vector<2xi1>
+    %8 = "vector.transfer_read"(%0, %2, %2, %2) <{"in_bounds" = [true], "operandSegmentSizes" = array<i32: 1, 2, 1, 0>, "permutation_map" = #map}> : (memref<4x4xindex>, index, index, index) -> vector<4xindex>
```
would it be worth adding custom syntax here? seems quite verbose

watermelonwolverine replied:
Sure, at some point it would totally be worth to add custom syntax. But I don't have the time at the moment because of my thesis..

superlopuh replied:
I'm happy to add it if you are. I think that a good way to merge this would be to first add the ops, and then add verification. I can open a new PR with your newly added operations with no verification at all, but with custom syntax, then this PR will track adding only the verifiers and verifier helpers. Does that sound OK?

watermelonwolverine replied:
Yes, sounds good

superlopuh replied:
No promises on when, Ill try to do it in the next week or two! Enjoy thesis writing ;)

superlopuh replied:
Hi @watermelonwolverine, do you have any estimates on when you'll have time to iterate on this? There's no urgent need for it, but it would be great to have an approximate date.

watermelonwolverine replied:
Hello. Sorry, I'm still super busy with my thesis and will be for a couple of weeks (at least 5). I can understand if somebody wants to take this over and finish it and I would be ok with it. Otherwise, I will continue working on this once I'm done with my master.

superlopuh replied:
No worries, this is super helpful, I'm happy to ping again in 5 weeks unless someone needs this urgently before then!

superlopuh replied:
Hello! Hope your thesis went well. Do you think you'll have some time in the next few weeks to iterate on this?

superlopuh replied:
Hello! I just found a use-case for these in my research today. Would you have time to add these over the next week? If not, do you have any objections to me taking over this PR?

superlopuh replied:
Monday or tuesday is great!

watermelonwolverine commented on xdsl/ir/affine/affine_expr.py:
```diff
@@ -308,6 +308,12 @@ def dfs(self) -> Iterator[AffineExpr]:
     def used_dims(self) -> set[int]:
         return {expr.position for expr in self.dfs() if isinstance(expr, AffineDimExpr)}
 
+    def is_function_of_dim(self, position: int) -> bool:
```
You were quicker than expected. I'm currently in the process of cleaning things up :sweat_smile:

watermelonwolverine commented on tests/filecheck/dialects/vector/vector_ops.mlir:
```diff
@@ -1,5 +1,5 @@
 // RUN: XDSL_ROUNDTRIP
-
+#map = affine_map<(d0, d1) -> (d0)>
```
Used in "permutation_map" of `vector.transfer_read`and `vector_transfer_write` line 13 and 14.

watermelonwolverine commented on xdsl/dialects/vector.py:
```diff
@@ -1076,6 +1076,8 @@ def parse(cls, parser: Parser) -> TransferWriteOp:
         InsertElementOp,
         TransferReadOp,
         TransferWriteOp,
+        TransferReadOp,
+        TransferWriteOp,
```
There are a few more problems. I'm working on it. I'll mark it as ready for review when I'm done

superlopuh commented on xdsl/dialects/vector.py:
```diff
@@ -846,6 +847,133 @@ def resolve_attrs(
 
         return resolved_mask, permutation_map, in_bounds
 
+    def is_broadcast_dim(self, dim: int) -> bool:
+        expr = self.permutation_map.data.results[dim]
+        if not isa(expr, AffineConstantExpr):
+            return False
+        return expr.value == 0
+
+    def has_broadcast_dim(self):
+        transfer_rank = len(self.permutation_map.data.results)
+
+        for dim in range(transfer_rank):
+            if self.is_broadcast_dim(dim):
+                return True
+
+        return False
```
I find this a little easier to read, and it avoids the whole "get range then index" pattern that's much slower in Python than in cpp.
```suggestion
    def has_broadcast_dim(self):
        """
        Return "true" if at least one of the vector dimensions is a broadcasted dimension.
        """
        return any(
            isinstance(expr, AffineConstantExpr) and expr.value == 0
            for expr in self.permutation_map.data.results
        )
```

watermelonwolverine commented on tests/filecheck/dialects/vector/vector_transfer_write_verify.mlir:
```diff
@@ -0,0 +1,50 @@
+
+%source, %vector, %index = "test.op"() : () -> (memref<?xf32>, vector<7xf32>, index)
+"vector.transfer_write"(%vector, %source, %index) <{in_bounds=[true], operandSegmentSizes = array<i32: 1, 1, 1, 0>, permutation_map = affine_map<(d0) -> (0)>}> : (vector<7xf32>, memref<?xf32>, index) -> ()
+// CHECK: should not have broadcast dimensions
```
There was a copy-paste  bug here where I checked for the line
```
CHECK: requires a permutation_map that is a permutation (found one dim used more than once)
```
Instead of 
```
// CHECK: should not have broadcast dimensions
```
However, it still succeeded.

I looked for ways too avoid this error but didn't find any. Do you know of a way to avoid this?
Using func.funcs and CHECK-LABEL could be a way but would require wrapping all tests cases with functions.

watermelonwolverine replied:
CHECK-LABEL did not work as expected because of the --verify-diagnostics flags. In the long run the best course may be to implement the --verify-diagnostics flag in xdsl-opt as it is implemented in mlir-opt.

watermelonwolverine commented on xdsl/dialects/vector.py:
```diff
@@ -846,6 +847,126 @@ def resolve_attrs(
+                f"{str(permutation_map)} vs in_bounds of of size {len(in_bounds)}"
+            )
+
+        for res, res_is_in_bounds in zip(
```
This check is not in the MLIR reference code (I assume it was there once, otherwise I don't think I would have added it).
There is also no test for this. Should I remove it?

watermelonwolverine commented on xdsl/dialects/vector.py:
```diff
@@ -846,6 +847,126 @@ def resolve_attrs(
+        ):
+            if isa(res, AffineConstantExpr) and not res_is_in_bounds:
+                raise VerifyException(
+                    f'"{op.name}" requires broadcast dimensions to be in-bounds'
```
This check is not in the MLIR reference code (I assume it was there once, otherwise I don't think I would have added it).
There is also no test for this. Should I remove it?

watermelonwolverine replied:
There is no verification check in MLIR and no test case in MLIR or xDSL that triggers this.

superlopuh replied:
Let's delete it then

superlopuh commented:
🚀
